src/lib.rs

```diff
@@ -19,25 +19,171 @@
 // IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 // CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
-//! BSON is a binary format in which zero or more key/value pairs are stored as a single entity.
-//! We call this entity a document.
+//! BSON, short for Binary JSON, is a binary-encoded serialization of JSON-like documents.
+//! Like JSON, BSON supports the embedding of documents and arrays within other documents
+//! and arrays. BSON also contains extensions that allow representation of data types that
+//! are not part of the JSON spec. For example, BSON has a datetime type and a binary data type.
 //!
-//! This library supports version 1.0 of the [BSON standard](http://bsonspec.org/spec.html).
+//! ```text
+//! // JSON equivalent
+//! {"hello": "world"}
 //!
-//! ## Basic usage
+//! // BSON encoding
+//! \x16\x00\x00\x00                   // total document size
+//! \x02                               // 0x02 = type String
+//! hello\x00                          // field name
+//! \x06\x00\x00\x00world\x00          // field value
+//! \x00                               // 0x00 = type EOO ('end of object')
+//! ```
+//!
+//! BSON is the primary data represenation for [MongoDB](https://www.mongodb.com/), and this crate is used in the
+//! [`mongodb`](https://docs.rs/mongodb/0.10.0/mongodb/) driver crate in its API and implementation.
+//!
+//! For more information about BSON itself, see [bsonspec.org](http://bsonspec.org).
+//!
+//! ## BSON values
+//!
+//! Many different types can be represented as a BSON value, including 32-bit and 64-bit signed
+//! integers, 64 bit floating point numbers, strings, datetimes, embedded documents, and more. To
+//! see a full list of possible BSON values, see the [BSON specification](http://bsonspec.org/spec.html). The various
+//! possible BSON values are modeled in this crate by the [`Bson`](enum.Bson.html) enum.
+//!
+//! ### Creating [`Bson`](enum.Bson.html) instances
+//!
+//! [`Bson`](enum.Bson.html) values can be instantiated directly or via the
+//! [`bson!`](macro.bson.html) macro:
+//!
+//! ```rust
+//! # use bson::{bson, Bson};
+//! let string = Bson::String("hello world".to_string());
+//! let int = Bson::I32(5);
+//! let array = Bson::Array(vec![Bson::I32(5), Bson::Boolean(false)]);
+//!
+//! let string: Bson = "hello world".into();
+//! let int: Bson = (5 as i32).into();
+//!
+//! let string: Bson = bson!("hello world");
+//! let int: Bson = bson!(5);
+//! let array = bson!([5, false]);
+//! ```
+//! [`bson!`](macro.bson.html) has supports both array and object literals, and it automatically
+//! converts any values specified to [`Bson`](enum.Bson.html), provided they are either `Serialize`
+//! or `Into<Bson>`.
+//!
+//! ### [`Bson`](enum.Bson.html) value unwrapping
+//!
+//! [`Bson`](enum.Bson.html) has a number of helper methods for accessing the underlying native Rust
```

Comment:
Does the macro actually accept things that are `Serialize` but not `Into<Bson>`? I'm not sure I've ever tried this.

Reply:
So it appears it actually doesn't. I was going off [this comment](https://github.com/mongodb/bson-rust/blob/master/src/macros.rs#L181) in the macro itself, but as you can see in the line below it only works on things that are `Into<Bson>`. Doing a little digging, it looks like the comment itself was lifted from [`serde_json`'s implementation of `json!`](https://github.com/serde-rs/json/blob/master/src/macros.rs#L270) way [back when](https://github.com/mongodb/bson-rust/pull/83/files#r142045639). I've updated this section to reflect that.

As a fun side note, this made me realize why `bson!` sometimes required a `Bson` type annotation (was calling `std::convert::From` in the macro instead of `Bson::from`), so I've fixed that.